### PR TITLE
[generate] Sdpa support softmax_offset

### PR DESCRIPTION
**Pull request by xiaoguoguo626807**
### PR Category

Execute Infrastructure

### PR Types

Improvements
### Description

sdpa 支持softmax_offset 参数

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 1]
```diff
@@ -83,6 +83,102 @@ def backward(ctx, d_scores):
...
+    # This is NOT equivalent to an additive mask (which softmax cancels out).
+    #
+    # Algorithm (mirrors flash_fwd.py / softmax.py finalize()):
+    #   scores  = Q @ K^T * scale                              [B, H, Q, K]
+    #   row_max = max(scores, dim=-1, keepdim=True)
+    #   exp_s   = exp(scores - row_max)
+    #   row_sum = sum(exp_s) + exp(sink_val - row_max)   ← virtual token
+    #   weights = exp_s / row_sum                        (sum < 1)
+    #   output  = weights @ V
+    scale_f = float(q_head_dim) ** -0.5
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py]: P1 **优先级：P1**

开启 `softmax_offset` 后会进入这个手写 SDPA 分支，但这里固定使用 `1 / sqrt(q_head_dim)`。`DotProductAttention` 的其它路径在 `_has_custom_softmax_scale` 为真时都会使用 `self.softmax_scale`，例如 MLA/YARN 的 mscale 或 `apply_query_key_layer_scaling`，下面的普通 SDPA 分支也会传 `scale=self.softmax_scale`。现在只要同时开启 softmax sink 和非默认 scale，QK logits 会按默认比例缩放，训练/推理结果会和 flashmask/普通 SDPA 路径不一致。

**处理要求：请针对该评论修复并提交新的 commit。**

建议把 scale 作为参数传入 helper，并从调用处传 `self.softmax_scale`，例如：

```python
def scaled_dot_product_attention_with_softmax_offset(..., scale=None):
    scale_f = float(scale if scale is not None else q_head_dim**-0.5)
...
scaled_dot_product_attention_with_softmax_offset(
    ...,
    q_head_dim=q_head_dim,
    scale=self.softmax_scale,
)
```

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 2]
```diff
@@ -83,6 +83,102 @@ def backward(ctx, d_scores):
...
+
+    row_max = scores.max(axis=-1, keepdim=True)  # [B, H, Q, 1]
+    exp_s = paddle.exp(scores - row_max)  # [B, H, Q, K]
+    row_sum = exp_s.sum(axis=-1, keepdim=True)  # [B, H, Q, 1]
+
+    # softmax_offset: [H] -> [1, Hq, 1, 1]
+    sink = softmax_offset.reshape([1, -1, 1, 1]).cast("float32")
+    row_sum = row_sum + paddle.exp(sink - row_max)  # add sink
+
+    weights = exp_s / row_sum  # [B, Hq, Q, K]
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py:172]: P1 **优先级：P1**

这个手写 `softmax_offset` 路径计算出 `weights` 后直接乘 `V`，没有像普通 SDPA 分支和 flashmask 分支一样应用 `self.config.attention_dropout`。因此只要训练配置里 `attention_dropout > 0` 且开启 softmax sink，这条新路径会完全关闭 attention dropout，改变正则化和训练行为。

**处理要求：请针对该评论修复并提交新的 commit。**

建议把 dropout 概率和 training 状态传入 helper，在 `weights` 之后、乘 `V` 之前应用：

```python
if dropout_p > 0.0:
    weights = paddle.nn.functional.dropout(
        weights,
        p=dropout_p,
        training=training,
    )
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现两个需要先修复的行为回归，细节已放在行级评论里：`softmax_offset` 的手写 SDPA 路径没有继承自定义 softmax scale，也没有应用 attention dropout。

另外，当前 CLA 检查仍为 pending：有一个提交作者显示为 `root`，需要在合入前让对应检查恢复通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/dot_product_attention.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/dot_product_attention.py:172]: P1 **优先级：P1**

复查当前 head 后，这个问题仍未修复：本次提交只给 helper 增加并传入了 `scale`，但 `scaled_dot_product_attention_with_softmax_offset` 仍没有 `dropout_p/training` 参数，`weights = exp_s / row_sum` 后仍直接进入 `matmul`。新增测试也只验证了分支路由，没有覆盖 `attention_dropout > 0` 时的语义。

**处理要求：请针对该评论修复并提交新的 commit。**

请继续按本线程里的建议，在 `weights` 乘 `V` 前按训练状态应用 attention dropout，并从调用处传入 `self.config.attention_dropout` 与 `self.training`。

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 3]
```diff
@@ -83,6 +83,103 @@ def backward(ctx, d_scores):
...
+        scores_g = paddle.matmul(q_g, k_g.transpose([0, 1, 2, 4, 3])) * scale_f
+        # [B, Hkv, groups, Q, K] → [B, Hq, Q, K]
+        scores = scores_g.reshape([bsz_cur, num_q_heads, q_len_cur, kv_len_cur])
+    else:
+        scores = (
+            paddle.matmul(q_f, k_f.transpose([0, 1, 3, 2])) * scale_f
+        )  # [B, Hq, Q, K]
+
+    if attn_mask_kv is not None:
+        scores = scores + attn_mask_kv.cast("float32")
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py]: P1 **优先级：P1**

这里把所有 `attn_mask_kv` 都 cast 成 float 后加到 logits，会破坏项目里常见的 bool mask 语义（`True` 表示 masked-out）。这种情况下 masked token 不会被屏蔽，反而会被加上 `1.0`，与原生 SDPA/其它 `attention_mask_func` 路径不一致。

**处理要求：请针对该评论修复并提交新的 commit。**

建议按 dtype 分支处理：bool mask 写成 `-inf`，只有 additive float mask 才做加法。

```suggestion
        if attn_mask_kv.dtype == paddle.bool:
            scores = paddle.where(
                attn_mask_kv,
                paddle.full_like(scores, float("-inf")),
                scores,
            )
        else:
            scores = scores + attn_mask_kv.cast("float32")
```

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 4]
```diff
@@ -83,6 +83,103 @@ def backward(ctx, d_scores):
...
+        )
+        scores = scores + neg_inf_mask
+
+    row_max = scores.max(axis=-1, keepdim=True)  # [B, H, Q, 1]
+    exp_s = paddle.exp(scores - row_max)  # [B, H, Q, K]
+    row_sum = exp_s.sum(axis=-1, keepdim=True)  # [B, H, Q, 1]
+
+    # softmax_offset: [H] -> [1, Hq, 1, 1]
+    sink = softmax_offset.reshape([1, -1, 1, 1]).cast("float32")
+    row_sum = row_sum + paddle.exp(sink - row_max)  # add sink
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py]: P1 **优先级：P1**

`row_max` 只来自真实 KV logits，没有把 sink logit 纳入同一个稳定化最大值。现有 sink reference 会先执行 `row_max = max(scores.max(...), sink)`；这里当 learnable sink 大于当前行最大分数时，`exp(sink - row_max)` 可能 overflow，真实 token 全被 mask 成 `-inf` 时也容易产生 NaN。

**处理要求：请针对该评论修复并提交新的 commit。**

```suggestion
    # softmax_offset: [H] -> [1, Hq, 1, 1]
    sink = softmax_offset.reshape([1, -1, 1, 1]).cast("float32")
    row_max = paddle.maximum(scores.max(axis=-1, keepdim=True), sink)
    exp_s = paddle.exp(scores - row_max)  # [B, H, Q, K]
    row_sum = exp_s.sum(axis=-1, keepdim=True) + paddle.exp(
        sink - row_max
    )
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查新提交：自定义 `softmax_scale` 的问题已经在当前代码中修复；我这边尝试关闭对应旧线程时权限不足，后续可由有权限的协作者处理。

仍有需要修复后再合入的问题：attention dropout 仍未应用，并且本次复查还补充了两个 `softmax_offset` 手写 SDPA 路径的正确性问题，细节见行级评论。

另外，当前 CLA 仍为 pending，合入前也需要恢复通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 5]
```diff
@@ -83,6 +83,110 @@ def backward(ctx, d_scores):
...
+        )
+        scores = scores + neg_inf_mask
+
+    # softmax_offset: [H] -> [1, Hq, 1, 1]
+    sink = softmax_offset.reshape([1, -1, 1, 1]).cast("float32")
+    row_max = paddle.maximum(scores.max(axis=-1, keepdim=True), sink)
+    exp_s = paddle.exp(scores - row_max)  # [B, H, Q, K]
+    row_sum = exp_s.sum(axis=-1, keepdim=True) + paddle.exp(
+        sink - row_max
+    )
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py]: P1 **优先级：P1**

当前 Codestyle Check 失败的原因是 `ruff format` 会重排这里的表达式。这个检查需要恢复通过后才能继续合入流程。

**处理要求：请针对该评论修复并提交新的 commit。**

```suggestion
    row_sum = exp_s.sum(axis=-1, keepdim=True) + paddle.exp(sink - row_max)
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交：bool mask 语义和 sink `row_max` 的两个问题已经修复。

仍有需要修复后再合入的问题：attention dropout 仍未应用（沿用已有行级线程），并且当前 Codestyle Check 失败，格式问题已补充为新的行级评论。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 6]
```diff
@@ -83,6 +83,118 @@ def backward(ctx, d_scores):
...
+    if attn_mask_kv is not None:
+        if attn_mask_kv.dtype == paddle.bool:
+            scores = paddle.where(
+                attn_mask_kv,
+                paddle.full_like(scores, float("-inf")),
+                scores,
+            )
+        else:
+            scores = scores + attn_mask_kv.cast("float32")
+    elif is_causal and query.shape[1] > 1:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py]: P1 **优先级：P1**

这里用 `elif` 会在 `attn_mask_kv` 非空时完全跳过 causal mask。当前调用方在普通训练/非 cache 路径会同时设置 `is_causal=True` 和 `attn_mask_kv=attention_mask`，原生 SDPA 会保留因果约束并叠加外部 mask；手写 `softmax_offset` 分支现在只要传入 padding/multimodal mask，就可能让 query 看到未来 key。

**处理要求：请针对该评论修复并提交新的 commit。**

```suggestion
    if is_causal and query.shape[1] > 1:
```

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 7]
```diff
@@ -83,6 +83,118 @@ def backward(ctx, d_scores):
...
+        q_len_cur, kv_len_cur = query.shape[1], key.shape[1]
+        causal_mask = paddle.tril(
+            paddle.ones([q_len_cur, kv_len_cur], dtype="float32")
+        )
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py:148]: P1 **优先级：P1**

这里的 causal mask 没有处理 `kv_len_cur > q_len_cur` 的 KV cache/prefill 场景。对于已经有历史 KV 的 query chunk，第一行 query 应该能看到前面的历史 key 和当前 token；现在 `tril([q_len, kv_len])` 只允许看到 key 0，会把大部分历史上下文错误屏蔽，和原生 SDPA 的 causal 语义不一致。

**处理要求：请针对该评论修复并提交新的 commit。**

```suggestion
        causal_mask = paddle.tril(
            paddle.ones([q_len_cur, kv_len_cur], dtype="float32"),
            diagonal=kv_len_cur - q_len_cur,
        )
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交：attention dropout 的修复已覆盖之前的问题；当前 Codestyle Check 仍失败，格式问题沿用已有行级线程。

这次复查还发现 `softmax_offset` 手写 SDPA 的 causal mask 与原生 SDPA 语义仍有差异，已在行级评论中说明，需要修复后再继续合入判断。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/dot_product_attention.py (hunk 7 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/dot_product_attention.py:148]: P1 **优先级：P1**

复查最新提交后，`elif` 已改为先应用 causal mask 再叠加 `attn_mask_kv`，但本线程的问题仍未修复：当前代码仍使用无 `diagonal` 的 `paddle.tril(paddle.ones([q_len_cur, kv_len_cur]))`。新加的 `test_causal_matches_manual_reference` 只覆盖 `q_len == kv_len == 3`，无法暴露 KV cache/prefill 中 `kv_len_cur > q_len_cur` 的偏移语义；例如 `q_len_cur=2, kv_len_cur=5` 时第一行 query 仍只会看到 key 0，而不是历史 key 0..2 加当前 key 3。

**处理要求：请针对该评论修复并提交新的 commit。**

建议继续按本线程的替换方式加入 offset，并补一个 `kv_len_cur > q_len_cur` 的回归测试：

```python
causal_mask = paddle.tril(
    paddle.ones([q_len_cur, kv_len_cur], dtype="float32"),
    diagonal=kv_len_cur - q_len_cur,
)
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交：attention dropout、外部 mask 与 causal mask 叠加、row_sum 格式，以及 vanilla SDPA 的 value padding 问题在当前代码中已修复；本地 `ruff format --diff` 对本 PR 两个文件也已通过。当前 CI 仍有构建任务在运行。

仍有一个阻塞点未完成：KV cache/prefill 场景下 `kv_len_cur > q_len_cur` 的 causal mask 偏移仍未修复，我已在原行级线程补充了最新证据。请修复该线程并补对应回归测试后再继续合入判断。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复查最新提交：`kv_len_cur > q_len_cur` 时的 causal mask 偏移已按前序意见修复，Nyanpasu 之前提出的阻塞项在当前 head 看起来已解决。

本地 `ruff format --diff` 对本 PR 两个文件通过；本地 pytest 因当前环境缺少 `paddle` 未能运行。当前已完成的 GitHub checks 为通过，剩余 CI 仍以最终结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 8]
```diff
@@ -507,22 +619,51 @@ def forward(
...
-                key,
-                value,
-                attn_mask_kv,
-                self.config.attention_dropout,
-                is_causal=is_causal,
-                **sdpa_kwargs,
-            )
+            # Handle MLA case where q/k head_dim != v head_dim (e.g. 192 vs 128)
+            # memory_efficient_attention does not support asymmetric head dims
+            sdpa_need_value_padding = q_head_dim != v_head_dim
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py:624]: P1 **优先级：P1**

这里把所有 `q_head_dim != v_head_dim` 都当成“把 V 补到 Q 维度”，但 `v_head_dim` 是可显式配置的，`DotProductAttention.forward` 也直接从 `value.shape[-1]` 取这个维度。若配置出现 `v_head_dim > q_head_dim`（例如 q/k head dim 为 32、V head dim 为 64），下一行会调用 `paddle.nn.functional.pad(value, [0, -32])`，这个负 padding 不是合法的补零路径，会让 softmax_offset 和 vanilla 两条 SDPA 分支在该配置下失败或得到错误形状。

**处理要求：请针对该评论修复并提交新的 commit。**

请把三种关系拆开处理，并补一个 `v_head_dim > q_head_dim` 的回归测试。示例结构如下；如果当前 SDPA kernel 不支持 `V` 维大于 `Q/K`，需要在该分支路由到显式 attention 实现，或至少在进入 `pad` 前给出清晰的 unsupported-config 错误，不能继续构造负 padding：

```python
if q_head_dim > v_head_dim:
    value_for_sdpa = paddle.nn.functional.pad(
        value, [0, q_head_dim - v_head_dim], value=0
    )
    sdpa_need_value_truncation = True
elif q_head_dim == v_head_dim:
    value_for_sdpa = value
    sdpa_need_value_truncation = False
else:
    # route to an implementation that supports Ev > E, or raise before pad()
    ...
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交：之前 Nyanpasu 提出的 causal mask、dropout、mask 叠加等问题在当前代码里仍保持修复状态；本轮 test-only 改动也改成了执行真实 attention 路径。

复查时发现 SDPA value padding 对 `v_head_dim > q_head_dim` 的配置仍会构造负 padding，已在行级评论中说明，需要修复后再继续合入判断。当前 CI 仍有多项任务运行中。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by changeyoung98 — APPROVED** [review]
